### PR TITLE
[11.0][ADD] account_fx_spot

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
Moved from https://github.com/OCA/account-financial-tools/pull/619

Account FX Spot
===============

This module extends the functionality of Accounting to support Foreign
Exchange Spot Transactions. A **Foreign Exchange Spot Transaction**, also
known as FX spot, is an agreement between two parties to buy one currency
against selling another currency at an agreed price for settlement on
the spot date. The exchange rate at which the transaction is done is
called the spot exchange rate.

Configuration
=============

To configure this module, you need to:

#. Go to *Accounting > Configuration > Settings*.
#. Enable the *Multi-Currencies* option.

Usage
=====

With this module, you will be able to:

* Create FX Spot Transactions.

#. Go to *Accounting > Foreign Exchange > FX Spot Transaction*.
#. Select a Partner and fill the outgoing and incoming amounts and currencies.

* Register Payments for FX Spot Transactions.

#. Select Several transactions from the tree view.
#. Click on *Action > Register Payment*.

or

#. Go to the form view for an open transaction.
#. Click on *Register Payment*.

### CODE REVIEW

**Review by JordiBForgeFlow — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@Yenthe666 you wanted this one, right? Your review is welcome!

File: account_fx_spot/models/account_fx_spot.py
Comment (Yenthe666) [inline account_fx_spot/models/account_fx_spot.py]: decorator?

File: account_fx_spot/models/account_fx_spot.py
Comment (Yenthe666) [inline account_fx_spot/models/account_fx_spot.py]: decorator?

**Comment by Yenthe666** [comment]
Hey @jbeficent,

Just a fast & highlevel review but it does look good.
P.S: do you guys never add the `.po` files?

**Comment by JordiBForgeFlow** [comment]
@Yenthe666 .po files are never added in OCA modules. Transifex is able to do the translations without them.

File: account_fx_spot/models/account_fx_spot.py
Comment (LoisRForgeFlow) [reply account_fx_spot/models/account_fx_spot.py]: api.multi is the default, right? but yeah maybe for consistency with the other methods in the file I should also explicitly set it for theese methods.

**Comment by JordiBForgeFlow** [comment]
Please squash your commits.

**Comment by LoisRForgeFlow** [comment]
@jbeficent I was waiting for @Yenthe666 to update his review. Diff could make that very quick for him!

File: account_fx_spot/models/account_fx_spot.py
Comment (Yenthe666) [reply account_fx_spot/models/account_fx_spot.py]: @lreficent yes it is the default. Some developers do not add it because it is 'the default', but what if at some point Odoo changes the API and this behaviour? Then you will have random functions breaking because of this.
I know this is a bug if and it feels like neatpicking but after all it only takes a few seconds to add an `@api.multi` right? 😉 
Thanks for adding!

**Review by Yenthe666 — APPROVED** [review]

**Comment by Yenthe666** [comment]
Thanks a lot, it does make it a lot easier. It looks good to me now, thanks for adding the decorators!

**Comment by JordiBForgeFlow** [comment]
@lreficent please squash and I will.merge.

File: account_fx_spot/models/account_fx_spot.py
Comment (LoisRForgeFlow) [reply account_fx_spot/models/account_fx_spot.py]: Yes, that's a fair point. Of course no problem to change it! Thanks

**Comment by LoisRForgeFlow** [comment]
@jbeficent Done!

**Comment by JordiBForgeFlow** [comment]
Did you check the compatibility with currency_rate_inverted?

**Review by JordiBForgeFlow — APPROVED** [review]

**Comment by feketemihai** [comment]
@lreficent Can you please adjust a little bit the view, seems a little bit strange and quite hard to use, by default you expect to input first the input amount and just after output amount,
 And i have an error when push save button:

The operation cannot be completed, probably due to the following:
- deletion: you may be trying to delete a record while other records still reference it
- creation/update: a mandatory field is not correctly set

[object with reference: out_currency_id - out.currency.id]

fxspot

**Comment by JordiBForgeFlow** [comment]
@feketemihai you have to provide both incoming and outgoing currency. Can you provide us the steps to reproduce the error?

**Comment by feketemihai** [comment]
@jbeficent i made the screenshot just for rearrange the fields. The error was with both currencies set, tested with currencies different than company currency, one equal to company currency, and the error was the same.

**Comment by LoisRForgeFlow** [comment]
@feketemihai I got the error, it is weird, I'm checking. 

About the screen, on my local the Incoming group was just at the right of the spot exchange group.

**Comment by feketemihai** [comment]
@lreficent The screenshot is from runbot.

File: account_fx_spot/views/account_fx_view.xml
Comment (LoisRForgeFlow) [inline account_fx_spot/views/account_fx_view.xml]: The issue are these divs where I use again in_currency_id and out_currency_id.

The desired effect is this:
image

What other options do I have? compute a string field to display it? any better idea?

**Comment by LoisRForgeFlow** [comment]
@feketemihai yes, I saw it on runbot too, let's first discuss the error, then the UI. See my comment above.

File: account_fx_spot/views/account_fx_view.xml
Comment (feketemihai) [reply account_fx_spot/views/account_fx_view.xml]: @lreficent Since rates are just for information i suggest to change those one to text and compute rate + in_currency + out_currency (and inverse), otherwise you have to add another fields for getting the currency names, so in total 4 fields, and to adjust tests after not to check float value of rate.

Comment (LoisRForgeFlow) [reply account_fx_spot/views/account_fx_view.xml]: just typed "outgoing" and "incoming", I think that works for now. I also switched the position of the incoming and outgoing groups as you suggested.

**Comment by feketemihai** [comment]
@lreficent Can you check, if i use different currencies that the payment journal, the outgoing payment is fine but for incoming it raises error that: Incoming Currencies does not match. It should be possible as well just like the outgoing...

**Comment by LoisRForgeFlow** [comment]
@feketemihai You are right, that is not supported at the moment but it would involve some tricky development into the account.payment logic. Plus, I don't think it would be the most common case in a FX spot transaction, specially for the incoming currency which is the one you are buying so you are supposed to have a specific bank/journal for it.

I added the issue in the README.

**Comment by feketemihai** [comment]
@lreficent Thanks for the issue and for this great module.

**Review by feketemihai — APPROVED** [review]

**Comment by feketemihai** [comment]
@pedrobaeza This one too, can be merged...Thanks.

**Comment by pedrobaeza** [comment]
@lreficent please squash before last 3 commits before merging

**Comment by LoisRForgeFlow** [comment]
@pedrobaeza Done!

**Pull request merged**
